File: java/leetcode086.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
	public ListNode partition(ListNode head, int x) {
		ListNode node1 = new ListNode(0), node2 = new ListNode(0);
		ListNode hoge = node1, fuga = node2;
		while (head != null) {
			if (head.val < x)
				hoge = hoge.next = head;
			else
				fuga = fuga.next = head;
			head = head.next;
		}
		fuga.next = null;
		hoge.next = node2.next;
		return node1.next;
	}
}

File: ruby/leetcode086.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @param {Integer} x
# @return {ListNode}
def partition(head, x)
	hoge = node1 = ListNode.new(0)
	fuga = node2 = ListNode.new(0)
	while head
		if head.val < x then
			hoge.next = head
			hoge = hoge.next
		else
			fuga.next = head
			fuga = fuga.next
		end
		head = head.next
	end
	fuga.next, hoge.next = nil, node2.next
	node1.next
end